src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java

```diff
@@ -0,0 +1,424 @@
+        this.solution = new TransientMAPFSolution();
+        this.agentPlans = new HashMap<>();
+        this.timeStamp = 0;
+
+        for (Agent agent : instance.agents) {
+            // init location of each agent to his source location
+            this.locations.put(agent, instance.map.getMapLocation(agent.source));
+
+            // init agents plans
+            this.agentPlans.put(agent, new SingleAgentPlan(agent, new ArrayList<Move>()));
+        }
+
+        // init agent's priority to unique number
+        initPriority(instance);
+
+        // init object who can find distance between every node in the grid to each agent's goal
+        this.heuristic = new DistanceTableAStarHeuristic(instance.agents, instance.map);
+    }
+
+
+    @Override
+    protected Solution runAlgorithm(MAPF_Instance instance, RunParameters parameters) {
+
+        // each iteration of the while represents timestamp
+        while (!(finished())) {
+
+            if (checkTimeout()) {
+                return null;
+            }
+
+            this.timeStamp++;
+
+            updatePriorities(instance);
+            // init agents that has not reached their goal
+            this.unhandledAgents = new ArrayList<>();
+            for (Map.Entry<Agent, Double> entry : this.priorities.entrySet()) {
+                Agent agent = entry.getKey();
+                Double priority = entry.getValue();
+                if (priority != -1.0) {
+                    // the agent did not reach his goal
+                    this.unhandledAgents.add(agent);
+                }
+            }
+
+            // nodes wanted in the next timestamp
+            this.takenNodes = new ArrayList<>();
+
+
+            while (!this.unhandledAgents.isEmpty()) {
+                Map.Entry<Agent, Double> maxEntry = getMaxEntry(this.priorities); // <agent, priority> pair with max priority
+                Agent cur = maxEntry.getKey(); // agent with the highest priority
+                solvePIBT(cur, null);
+            }
+
+            // if agent reached his goal, we don't add him to this.unhandledAgents
+            // so, iterate on all agents and add Move of stay in place if "solvePIBT" call didn't move the agent
+            if (!(finished())) {
+                for (Map.Entry<Agent, Double> entry : this.priorities.entrySet()) {
+                    Agent agent = entry.getKey();
+                    Double priority = entry.getValue();
+                    // the agent reached his goal
+                    // add new move to the agent's plan - stay in current node
+                    if (priority == -1.0 && this.timeStamp - this.agentPlans.get(agent).size() == 1) {
+                        Move move = new Move(agent, this.timeStamp, this.locations.get(agent), this.locations.get(agent));
+                        if (this.constraints.accepts(move)) {
+                            this.agentPlans.get(agent).addMove(move);
+                        }
```

Comment:
Looks like this might cause problems because the plan would just have holes.
```suggestion
//                        if (this.constraints.accepts(move)) {
//                            this.agentPlans.get(agent).addMove(move);
//                        } // TODO what should an agent do in this situation?
```

Reply:
i added else statement : "solvePIBT(agent, null);"

Reply:
I'm not sure how that works but I trust you. Just add a test that covers this code path